Allow attaching files to TestResults

ChrisMaddock opened the pull request:
This is the framework functionality required for #1670. (Adding this PR to the milestone separately, as I want to track runner implementation with the issue.)

Specific design points:

- API is `TestContext.AddTestAttachment(string filepath, string description = null)` Are we happy with that?

- I forced all files to be absolute paths - based on the logic that it's the runner that decides where the test result XML will end up, I believe? This is the same approach MSTest takes.

What do people think? I've also added some line comments on things to check.

ChrisMaddock commented on src/NUnitFramework/framework/Internal/TestExecutionContext.cs:
```diff
@@ -493,6 +502,28 @@ public void AddFormatter(ValueFormatterFactory formatterFactory)
+
+            if (!Path.IsPathRooted(filepath))
+                filepath = Path.Combine(TestContext.CurrentContext.WorkDirectory, filepath);
+
```
Is this the correct way to build an absolute file path? I wanted to use `Path.GetFullPath()` - but it's not available in PCL.

rprouse replied:
Looks good to me, although I think `Path.GetFullPath()` will be available once I get the .NET Standard version working.

CharliePoole replied:
Actually `Path.GetFullPath()` would be incorrect. We promise to store all output files relative to our WorkDirectory, which is __not__ the current directory. The choice of name for the WorkDirectory, may have been a bad one, since it sounds a bit like the working directory, i.e. cwd, but I believe we are stuck with it.

A more basic question... why do we want to make this absolute at all? Why not just use what the user gives us?

ChrisMaddock replied:
Thanks for the explanation! 

This was something else I went back and forth on. I wanted to make it absolute, as I had previously thought:
- The attachments should always be relative the the results xml - as the file paths written there should be correct to the files current location.
- The runner controls where the results xml is written, meaning that the runner could theoretically move where the results ended up.

What I didn't realise is that WorkDirectory wasn't the working directory. 😄 So will any results xml file always end up in TestContext.WorkDirectory? Or is that a runner setting, could TestResults be written anywhere, like I'd first thought?

I did think full paths would be easier to resolve should the user move the result xml file around for any CI etc. Is there a disadvantage to writing full paths, if we already define that they're relative to WorkDirectory? I guess if they remain relative, and you were adding all your attachments to the same subdir, you could move the directory as a whole?

CharliePoole replied:
The runner passes Work directory to the framework to be used in exactly this circumstance. The XML result file is saved there. 
In the case of the console runner, it can be set by a command-line argument.

ChrisMaddock replied:
Ok, cool. I'll remove the stuff that meddles with paths in that case - if we have the concrete definition that paths will always be relative to WorkDirectory/results location. Thanks.

ChrisMaddock commented on src/NUnitFramework/tests/nunit.framework.tests-3.5.csproj:
```diff
@@ -122,6 +123,7 @@
     <Compile Include="Internal\Results\TestResultOutputTests.cs" />
     <Compile Include="Internal\Results\TestResultSuccessTests.cs" />
     <Compile Include="Internal\Results\TestResultTests.cs" />
+    <Compile Include="Internal\Results\TestResultWarningTests.cs" />
```
This is unrelated - just noticed this file was missing from all except the 2.0 build.

ChrisMaddock commented:
Regarding file location - MSTest also copies the file - but has a fully fledged directory structure to support that.

This currently just references whatever the user passes in - meaning the link will be broken if the results are transferred between machines. Do we want to go any further, and (e.g.) copy the files to a directory next to the result xml? I wasn't entirely sure of the advantage, given a) potential copy permission errors and b) still requiring the user to copy both files and results, if that was the case.

I'm tempted to think the simple option will be ok, until such a point as someone has a compelling argument to change it?

CharliePoole commented:
@ChrisMaddock What do you mean by "I want to track runner implementation with the issue?"

ChrisMaddock commented:
@CharliePoole - There's a number of people following issue #1670. Rather than close it, and leave them to track down the related runner issues, which need to be completed before they can use the functionality, I've converted it to an Epic, and will add issues for the different runners, once we've decided what to do for each.

The side effect of that, is that there's no dedicated issue for just the framework component of #1670 - so I added the PR to the Epic.

CharliePoole commented:
I'm glad you took this on as it's a needed feature. I do need a bit of general explanation before doing a detailed review.

You are adding an entry to TestExecutionContext for attachments and then copying them to the result. Is there any reason you don't want to simply add them to the result directly? This would be analogous to how we add text output to the result.

In the TestExecutionContext constructor, you are copying the attachments from each higher level to the lower level. Is there a specific reason for that? Won't that mean the attachments are duplicated all the way down the tree of results?

ChrisMaddock commented:
> You are adding an entry to TestExecutionContext for attachments and then copying them to the result. Is there any reason you don't want to simply add them to the result directly? This would be analogous to how we add text output to the result.

I hadn't realised the TestResult was created at the start of the test - I followed where StartTime was set, which was in WorkItem.WorkItemComplete(). This sounds like a much better idea.

> In the TestExecutionContext constructor, you are copying the attachments from each higher level to the lower level. Is there a specific reason for that? Won't that mean the attachments are duplicated all the way down the tree of results?

Nope - that's just me not looking into things properly. Sorry.

It's getting on a bit UK-time, will make the changes tomorrow. Thanks for those!

CharliePoole commented:
@ChrisMaddock Great! I'll wait till you update to do a detailed review.

ChrisMaddock commented on src/NUnitFramework/tests/Internal/TestExecutionContextTests.cs:
```diff
@@ -907,6 +918,65 @@ public void SetAndRestoreValueFormatter()
+            var context = new TestExecutionContext { CurrentResult = fakeResult };
+            return context;
+        }
+
```
`CreateFakeContext()` is pretty ugly. Am I missing a better way to create temporary context to test with? I didn't want to attach anything to the actual TestExecutionContext.

CharliePoole replied:
It's easy to create a context, harder to populate it with meaningful values. In this case, we needed a context with a valid result. The alternative is to use NSubstitute but I think you'll be doing a lot of setup if you do that.

ChrisMaddock commented:
Initial changes made, and ready for review. Thanks Charlie.

jnm2 commented on src/NUnitFramework/framework/Internal/Results/TestResult.cs:
```diff
@@ -268,7 +283,7 @@ internal set
             }
         }
-
+        
```
I really hate to nitpick whitespace. I like these tools:
VS: https://marketplace.visualstudio.com/items?itemName=MadsKristensen.TrailingWhitespaceVisualizer
VS Code: https://marketplace.visualstudio.com/items?itemName=shardulm94.trailing-spaces with `"trailing-spaces.trimOnSave": true`

rprouse replied:
@jnm2, LOL, nice to see other developers as OCD as me 😄

CharliePoole replied:
OTOH, if VS doesn't really care...

jnm2 commented on src/NUnitFramework/framework/Internal/Results/TestResult.cs:
```diff
@@ -701,6 +719,26 @@ private ResultState AssertionStatusToResultState(AssertionStatus status)
+        /// </summary>
+        /// <param name="targetNode">The target node.</param>
+        /// <returns>The new attachments element.</returns>
+        private TNode AddAttatchmentsElement(TNode targetNode)
```
Nit: should be `AddAttachmentsElement`

ChrisMaddock replied:
I never could spell that word! Thanks!

CharliePoole replied:
Spelling is not a nitpick. Just whitespace. 😄

jnm2 commented on src/NUnitFramework/framework/Internal/TestAttachment.cs:
```diff
@@ -0,0 +1,53 @@
+        /// <summary>
+        /// Absolute filepath to attachment file
+        /// </summary>
+        public string Filepath { get; }
```
I've never seen this as one word. Can it be `FilePath`?

ChrisMaddock replied:
I Googled, and you're right. I'll change throughout - thanks!

jnm2 commented on src/NUnitFramework/framework/Internal/TestAttachment.cs:
```diff
@@ -0,0 +1,53 @@
+        /// </summary>
+        /// <param name="filepath">Absolute filepath to attachment file</param>
+        /// <param name="description">Optional user specifed description of attachment</param>
+        public TestAttachment(string filepath, string description)
```
Same; I've never seen this as one word. Can it be `filePath`?

jnm2 commented on src/NUnitFramework/framework/Internal/TestExecutionContext.cs:
```diff
@@ -493,6 +494,28 @@ public void AddFormatter(ValueFormatterFactory formatterFactory)
+        /// </summary>
+        /// <param name="filepath">Filepath to attachment. Can be absolute, or relative to working directory.</param>
+        /// <param name="description">Optional description of attachment</param>
+        public void AddTestAttachment(string filepath, string description = null)
```
Same; I've never seen this as one word. Can it be filePath?

jnm2 commented on src/NUnitFramework/tests/Internal/Results/TestResultAttachmentTests.cs:
```diff
@@ -0,0 +1,106 @@
+        [SetUp]
+        public void SetUp()
+        {
+            _result = TestUtilities.Fakes.GetTestMethod(this, "FakeMethod").MakeTestResult();
```
`nameof(FakeMethod)` 😀

jnm2 commented on src/NUnitFramework/tests/Internal/TestExecutionContextTests.cs:
```diff
@@ -43,28 +44,38 @@ namespace NUnit.Framework.Internal
         {
             _fixtureContext = TestExecutionContext.CurrentContext;
             _fixtureResult = _fixtureContext.CurrentResult.ResultState;
+            _tempFilePath = Path.Combine(TestContext.CurrentContext.WorkDirectory, "NUnitTests.tmp");
```
Replace `"NUnitTests.tmp"` with `TempFileName`?

ChrisMaddock replied:
Thanks. I had to do a bit of shuffling here to get things playing nice with PCL - must have missed that. 🙂

jnm2 commented on src/NUnitFramework/tests/Internal/TestExecutionContextTests.cs:
```diff
@@ -907,6 +918,65 @@ public void SetAndRestoreValueFormatter()
+
+        public TestExecutionContext CreateFakeContext()
+        {
+            var fakeResult = TestUtilities.Fakes.GetTestMethod(this, "FakeMethod").MakeTestResult();
```
`nameof(FakeMethod)` 😀

ChrisMaddock commented:
Thanks @jnm2. I'll wait for comments from others, then fix everything in one go. 😄

rprouse commented on src/NUnitFramework/framework/Internal/Results/TestResult.cs:
```diff
@@ -701,6 +719,26 @@ private ResultState AssertionStatusToResultState(AssertionStatus status)
+                var attachmentNode = attachmentsNode.AddElement("attachment", attachment.Filepath);
+
+                if (attachment.Description != null)
+                    attachmentNode.AddAttribute("description", attachment.Description);
```
Description is a user supplied string that may be long or require CDATA encoding. Based on that, we should probably make it a child element rather than an attribute and CDATA encode it.

```C#
var attachmentNode = attachmentsNode.AddElement("attachment");
attachmentNode.AddElement("filename", attachment.Filepath);

if (attachment.Description != null)
    attachmentNode.AddElementWithCDATA("description", attachment.Description);
```

Tests will need to be changed accordingly.

ChrisMaddock replied:
Sure. 

I did consider this - the current AddAttribute goes through an ExcapeXmlCharacters() - which should sufficiently encode any user string. If length is a concern for readability however, I can change it over. Will do that with the rest. 🙂

CharliePoole replied:
Amazing how things grow! I didn't imagine we would have a description at all at the start but if we have one I think @rprouse is right.

ChrisMaddock commented:
Thanks both. @CharliePoole - you said earlier you were planning to come back to this - so I'll hold on for your feedback, then deal with everything in one go.

CharliePoole commented on src/NUnitFramework/framework/Internal/Results/TestResult.cs:
```diff
@@ -80,7 +80,13 @@ public abstract class TestResult : ITestResult
+        /// TestAttachments attached to result.
+        /// Only exposed for testing.
+        /// </summary>
+        internal ICollection<TestAttachment> TestAttachments { get; } = new List<TestAttachment>();
```
Working my way down... I can't see why this would not be `private readonly List`. Maybe it will become clear as I go. 😄

jnm2 replied:
Comment says "Only exposed for testing." but `readonly` sounds good.

CharliePoole replied:
Ah! I didn't notice the comment. So this field is not the private backing field for a public property then? That seems a bit odd, since it leaves a part of the TestResult inaccessible to clients.

ChrisMaddock replied:
You're right. I had been keeping this as an 'add-only' collection, but it would be more in-keeping with the rest of the TestResult class if this was exposed to the user in a ReadOnlyCollection. I'll expose this list and the TestAttachment type.

CharliePoole commented on src/NUnitFramework/framework/Internal/TestExecutionContext.cs:
```diff
@@ -493,6 +494,28 @@ public void AddFormatter(ValueFormatterFactory formatterFactory)
+
+#if !PORTABLE
+            if (!File.Exists(filepath))
+                throw new FileNotFoundException("Test attachment filepath could not be found.", filepath);
```
Why do we care that the file doesn't exist? Maybe the user asked to add the attachment before saving it. Maybe some third party software is expected to add it later. If we are not doing anything with the file, why not just record what the user tells us to record?

ChrisMaddock replied:
This was a point I went back and forth on. 🙂 My thinking was that adding a non-existant file is far more likely to occur through user-error, rather than intention, and you'd want to know about that ASAP. I imagine a common use case may be where people only refer to attachments when the test has failed - so broken paths may only be discovered 6 months after writing, when the test first fails, and it's too late.

The opposite arguments e.g. 'adding an attachment before saving it', mostly seem like things that could be managed by the user. (e.g. Don't add an attachment before you've created it!) That seems like fairly bad practice anyway, what if the creation of the attachment later went wrong, and the file couldn't be written? We could always add a 'don't check file existence' overload in the future if such cases arose.

Not that it matters, but MSTest checks if the File Exists. @rprouse @jnm2 - what do you guys think?

CharliePoole replied:
I guess checking for existence makes sense in general.

jnm2 replied:
Typically, if you're going to do anything else with the file other than verify that it exists, the existence check is meaningless because it's race condition prone. To have meaningful handling you'd have to catch FileNotFoundException because it's the only atomic API.

> I imagine a common use case may be where people only refer to attachments when the test has failed - so broken paths may only be discovered 6 months after writing, when the test first fails, and it's too late.

But this is a great point and would be very helpful.

CharliePoole commented on src/NUnitFramework/framework/TestContext.cs:
```diff
@@ -295,6 +295,16 @@ public static void AddFormatter(ValueFormatterFactory formatterFactory)
+        /// <param name="description">Optional description of attachment</param>
+        public static void AddTestAttachment(string filepath, string description = null)
+        {
+            TestExecutionContext.CurrentContext.AddTestAttachment(filepath, description);
```
Why the extra level of indirection? In other code, TestContext just accesses `TestExecutionContext.CurrentContext.CurrentResult`  directly.

ChrisMaddock replied:
I can change that if this ways preferred. I get a little mixed up about logic should live between TestContext and TestExecutionContext - I'd followed how `AddFormatter()` was written.

This is also more possible now I'm no longer storing an intermediate list of attachments on TestExecutionContext. 😄

ChrisMaddock commented:
Thanks @CharliePoole! I've replied to the inline comments, the remaining ones...

> I would have tested this using a new class in the testdata-assembly. That would eliminate all worries about test context, execution context, etc. We would simply be looking at the result and it's fields.

I'll try it. 🙂 

> There is no test of TestContext
No - I'll move the tests currently in TestExecutionContext to use TestContext, once I move everything out of there.

> There are no new tests of TestResult, particularly to verify that it creates XML correctly.
Is this not covered by TestResultAttachmentTests?

Will hopefully make the changes this weekend - thanks all!

ChrisMaddock commented:
Review changes made - this is ready for a second look. 🙂 

Only thing to note, is that this contains `if #PORTABLE` symbols, so either this or #2174 will need adjusting, depending which is merged first.

oznetmaster commented:
I commented on the commit.  Probably the wrong place, again :(
